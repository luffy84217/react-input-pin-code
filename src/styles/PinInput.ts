import styled from 'styled-components';

export const PinInputContainer = styled.div`
  display: flex;
  align-items: center;
`;
