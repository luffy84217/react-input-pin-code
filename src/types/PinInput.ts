import React from 'react';

export type PinInputProps = {
  values: string[];
  type?: 'number' | 'text';
  mask?: boolean;
  validate?: string | string[] | RegExp;
  showState?: boolean;
  onChange?: (
    value: string | string[],
    index: number,
    values: string[]
  ) => void;
  onComplete?: (values: string[]) => void;
} & Pick<React.InputHTMLAttributes<HTMLInputElement>, 'inputMode'>;
