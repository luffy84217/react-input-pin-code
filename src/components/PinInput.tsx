import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { PinInputContainer } from '../styles/PinInput';
import { PinInputProps } from '../types/PinInput';
import PinInputField from './PinInputField';
import { validateToPattern } from '../utils';

const propTypes = {
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
  type: PropTypes.oneOf<'number' | 'text'>(['number', 'text']),
  mask: PropTypes.bool,
  validate: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.instanceOf(RegExp),
  ]),
  showState: PropTypes.bool,
  onChange: PropTypes.func,
};

const PinInput: React.FC<PinInputProps> = (props) => {
  const completed = useMemo(
    () => props.values.every((val) => val),
    [props.values]
  );

  if (completed && props.onComplete) {
    if (props.validate) {
      if (
        props.values.every((val) =>
          new RegExp(validateToPattern(props.validate)).test(val)
        )
      ) {
        props.onComplete(props.values);
      }
    } else {
      props.onComplete(props.values);
    }
  }

  return (
    <PinInputContainer>
      {props.values.map((value, i) => (
        <PinInputField
          key={i}
          index={i}
          value={value}
          completed={completed}
          {...props}
        />
      ))}
    </PinInputContainer>
  );
};

PinInput.displayName = 'PinInput';
PinInput.propTypes = propTypes;
PinInput.defaultProps = {
  type: 'number',
  mask: false,
  showState: true,
};

export default PinInput;
