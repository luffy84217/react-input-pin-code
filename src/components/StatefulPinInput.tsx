import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { StatefulPinInputProps } from '../types/StatefulPinInput';
import PinInput from './PinInput';

const propTypes = {
  length: PropTypes.number,
  initialValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  type: PropTypes.oneOf<'number' | 'text'>(['number', 'text']),
  mask: PropTypes.bool,
  validate: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.instanceOf(RegExp),
  ]),
  format: PropTypes.func,
  showState: PropTypes.bool,
  size: PropTypes.oneOf<'xs' | 'sm' | 'md' | 'lg'>(['xs', 'sm', 'md', 'lg']),
  autoFocus: PropTypes.bool,
  autoTab: PropTypes.bool,
  containerStyle: PropTypes.object,
  inputStyle: PropTypes.object,
  onChange: PropTypes.func,
};

const StatefulPinInput: React.FC<StatefulPinInputProps> = ({
  length,
  initialValue,
  onChange,
  ...props
}) => {
  const [values, setValues] = useState(
    Array.from({ length }, (_v, i) => initialValue[i] || '')
  );

  const handleInputChange = (
    value: string | string[],
    index: number,
    values: string[]
  ) => {
    setValues(values);
    if (onChange) {
      onChange(value, index, values);
    }
  };

  return <PinInput values={values} onChange={handleInputChange} {...props} />;
};

StatefulPinInput.displayName = 'StatefulPinInput';
StatefulPinInput.propTypes = propTypes;
StatefulPinInput.defaultProps = {
  length: 4,
  initialValue: '',
  type: 'number',
  mask: false,
  showState: true,
  size: 'md',
  autoFocus: false,
  autoTab: true,
  containerStyle: {},
  inputStyle: {},
  autoComplete: 'off',
  placeholder: 'o',
  'aria-label': 'Please enter pin code',
};

export default StatefulPinInput;
