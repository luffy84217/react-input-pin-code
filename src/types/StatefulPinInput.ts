import { PinInputProps } from './PinInput';

export type StatefulPinInputProps = Omit<PinInputProps, 'values'> & {
  length?: number;
  initialValue?: string | string[];
};
