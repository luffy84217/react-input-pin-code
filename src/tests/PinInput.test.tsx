import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { fireEvent, render } from '@testing-library/react';
import PinInput from '../components/PinInput';
import StatefulPinInput from '../components/StatefulPinInput';

describe('Pin Input', () => {
  it('should contain 4 inputs', () => {
    const { getAllByRole } = render(<PinInput values={['', '', '', '']} />);

    const PinInputFields = getAllByRole('textbox');

    expect(PinInputFields).toHaveLength(4);
  });

  it('should call complete callback when filling all inputs', () => {
    let mockValues = ['1', '2', '3', ''];
    const mockCompleteHandler = jest.fn();
    const { rerender } = render(
      <PinInput values={['1', '2', '3', '']} onComplete={mockCompleteHandler} />
    );

    expect(mockCompleteHandler).not.toBeCalled();

    mockValues = ['1', '2', '3', '4'];

    rerender(<PinInput values={mockValues} onComplete={mockCompleteHandler} />);

    expect(mockCompleteHandler).toBeCalledTimes(1);
    expect(mockCompleteHandler).toBeCalledWith(mockValues);
  });

  it('should only call complete callback when filling all inputs and passing validation if it exists', () => {
    let mockValues = ['1', '2', '3', 'a'];
    const pattern = ['1', '2', '3', '4'];
    const mockCompleteHandler = jest.fn();
    const { rerender } = render(
      <PinInput
        values={mockValues}
        validate={pattern}
        onComplete={mockCompleteHandler}
      />
    );

    expect(mockCompleteHandler).not.toBeCalled();

    mockValues = ['1', '2', '3', '4'];

    rerender(
      <PinInput
        values={mockValues}
        validate={pattern}
        onComplete={mockCompleteHandler}
      />
    );

    expect(mockCompleteHandler).toBeCalledTimes(1);
    expect(mockCompleteHandler).toBeCalledWith(mockValues);
  });

  it('should render correct id if it exists', () => {
    const mockId = 'foo';
    const { getAllByRole } = render(
      <PinInput values={['', '', '', '']} id={mockId} />
    );

    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    PinInputFields.forEach((input, i) => {
      expect(input.id).toBe(`${mockId}-${i}`);
    });
  });

  it('should auto focus first input if autoFocus is true', () => {
    const { getAllByRole } = render(
      <PinInput values={['', '', '', '']} autoFocus />
    );

    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    expect(PinInputFields[0]).toBe(document.activeElement);
  });

  it('should call focus and blur callback if they exist', () => {
    const mockKeydownHandler = jest.fn();
    const mockFocusHandler = jest.fn();
    const mockBlurHandler = jest.fn();
    const { getAllByRole } = render(
      <PinInput
        values={['', '', '', '']}
        onKeyDown={mockKeydownHandler}
        onFocus={mockFocusHandler}
        onBlur={mockBlurHandler}
      />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    expect(mockKeydownHandler).not.toBeCalled();
    expect(mockFocusHandler).not.toBeCalled();
    expect(mockBlurHandler).not.toBeCalled();

    PinInputFields[1].focus();
    fireEvent.keyDown(PinInputFields[1], {
      key: 'Backspace',
      code: 'Backspace',
    });
    PinInputFields[0].blur();

    expect(mockKeydownHandler).toBeCalled();
    expect(mockFocusHandler).toBeCalled();
    expect(mockBlurHandler).toBeCalled();
  });

  it('should auto focus previous input if pressing backspace when current input is empty', () => {
    const mockKeydownHandler = jest.fn();
    const { getAllByRole } = render(
      <PinInput values={['', '', '', '']} onKeyDown={mockKeydownHandler} />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    expect(PinInputFields[0]).not.toBe(document.activeElement);
    expect(mockKeydownHandler).not.toBeCalled();

    fireEvent.click(PinInputFields[1]);
    fireEvent.keyDown(PinInputFields[1], {
      key: 'Backspace',
      code: 'Backspace',
    });

    expect(PinInputFields[0]).toBe(document.activeElement);
    expect(mockKeydownHandler).toBeCalled();
  });

  it('should replacing each character with a symbol if mask is true', () => {
    const { getAllByPlaceholderText } = render(
      <PinInput values={['', '', '', '']} mask={true} type="text" />
    );
    const PinInputFields = getAllByPlaceholderText('o') as HTMLInputElement[];

    PinInputFields.forEach((input) => {
      expect(input.type).toBe('password');
    });
  });

  it('should render correct style when size is provided', () => {
    const { getAllByRole, rerender } = render(
      <PinInput values={['', '', '', '']} size="xs" />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];
    const xsTree = renderer
      .create(<PinInput values={['', '', '', '']} size="xs" />)
      .toJSON();
    const smTree = renderer
      .create(<PinInput values={['', '', '', '']} size="sm" />)
      .toJSON();
    const mdTree = renderer
      .create(<PinInput values={['', '', '', '']} size="md" />)
      .toJSON();
    const lgTree = renderer
      .create(<PinInput values={['', '', '', '']} size="lg" />)
      .toJSON();

    expect(xsTree).toMatchSnapshot();
    expect(smTree).toMatchSnapshot();
    expect(mdTree).toMatchSnapshot();
    expect(lgTree).toMatchSnapshot();

    PinInputFields.forEach((input) => {
      expect(input).toHaveStyle({ width: '1.5rem' });
      expect(input).toHaveStyle({ height: '1.5rem' });
    });
    rerender(<PinInput values={['', '', '', '']} size="sm" />);
    PinInputFields.forEach((input) => {
      expect(input).toHaveStyle({ width: '2rem' });
      expect(input).toHaveStyle({ height: '2rem' });
    });
    rerender(<PinInput values={['', '', '', '']} size="md" />);
    PinInputFields.forEach((input) => {
      expect(input).toHaveStyle({ width: '2.5rem' });
      expect(input).toHaveStyle({ height: '2.5rem' });
    });
    rerender(<PinInput values={['', '', '', '']} size="lg" />);
    PinInputFields.forEach((input) => {
      expect(input).toHaveStyle({ width: '3rem' });
      expect(input).toHaveStyle({ height: '3rem' });
    });
  });

  it('should render correct style when border color is provided', () => {
    const { getAllByRole, rerender } = render(
      <PinInput values={['1', '2', '3', '4']} validBorderColor="#198754" />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];
    const validTree = renderer
      .create(
        <PinInput values={['1', '2', '3', '4']} validBorderColor="#198754" />
      )
      .toJSON();

    expect(validTree).toMatchSnapshot();
    expect(PinInputFields[0]).toHaveStyle('border-color: rgb(25,135,84)');

    rerender(
      <PinInput
        values={['a', 'b', 'c', 'd']}
        validate="abc"
        errorBorderColor="#dc3545"
      />
    );
    const invalidTree = renderer
      .create(
        <PinInput
          values={['a', 'b', 'c', 'd']}
          validate="abc"
          errorBorderColor="#dc3545"
        />
      )
      .toJSON();

    expect(invalidTree).toMatchSnapshot();

    expect(PinInputFields[3]).toHaveStyle('border-color: rgb(220,53,69)');
  });

  it('should render correct when showState is false', () => {
    const { getAllByRole } = render(
      <PinInput values={['1', '2', '3', '4']} showState={false} />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];
    const tree = renderer
      .create(<PinInput values={['1', '2', '3', '4']} showState={false} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(PinInputFields[0]).toHaveStyle('border-color: #cccccc');
  });

  it('should auto fill when input is mulitple characters', () => {
    const { getAllByRole, rerender } = render(<StatefulPinInput />);
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    fireEvent.change(PinInputFields[0], { target: { value: '0123' } });
    PinInputFields.forEach((input, i) => {
      expect(input.value).toBe(i.toString());
    });

    rerender(<StatefulPinInput initialValue={['0', '', '', '']} />);
    fireEvent.change(PinInputFields[0], { target: { value: '0123' } });
    PinInputFields.forEach((input, i) => {
      expect(input.value).toBe(i.toString());
    });
  });

  it('should delete character if Backspace is pressed', () => {
    const { getAllByRole } = render(
      <StatefulPinInput initialValue={['0', '', '', '']} />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    fireEvent.change(PinInputFields[0], { target: { value: '' } });
    expect(PinInputFields[0].value).toBe('');
  });

  it('should override the input value with the last character typed', () => {
    const { getAllByRole } = render(
      <StatefulPinInput initialValue={['0', '', '', '']} />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    fireEvent.change(PinInputFields[0], { target: { value: '01' } });
    expect(PinInputFields[0].value).toBe('1');
    fireEvent.change(PinInputFields[0], { target: { value: '21' } });
    expect(PinInputFields[0].value).toBe('2');
  });

  it('should format input value', () => {
    const { getAllByRole } = render(
      <StatefulPinInput
        type="text"
        format={(c) => (parseInt(c) + 1).toString()}
      />
    );
    const PinInputFields = getAllByRole('textbox') as HTMLInputElement[];

    fireEvent.change(PinInputFields[0], { target: { value: '1' } });
    expect(PinInputFields[0].value).toBe('2');
    fireEvent.change(PinInputFields[1], { target: { value: '23' } });
    expect(PinInputFields[1].value).toBe('3');
    expect(PinInputFields[2].value).toBe('4');
  });
});
