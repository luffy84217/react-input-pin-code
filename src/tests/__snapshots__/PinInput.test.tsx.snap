// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pin Input should render correct style when border color is provided 1`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 2.5rem;
  height: 2.5rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 1rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:valid {
  border-color: rgb(25,135,84);
  box-shadow: rgb(25,135,84) 0px 0px 0px 1px;
  background-color: rgba(25,135,84,0.1);
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="1"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="2"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="3"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="4"
  />
</div>
`;

exports[`Pin Input should render correct style when border color is provided 2`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 2.5rem;
  height: 2.5rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 1rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:valid {
  border-color: rgb(25,135,84);
  box-shadow: rgb(25,135,84) 0px 0px 0px 1px;
  background-color: rgba(25,135,84,0.1);
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    pattern="a|b|c"
    placeholder="o"
    style={Object {}}
    type="text"
    value="a"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    pattern="a|b|c"
    placeholder="o"
    style={Object {}}
    type="text"
    value="b"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    pattern="a|b|c"
    placeholder="o"
    style={Object {}}
    type="text"
    value="c"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    pattern="a|b|c"
    placeholder="o"
    style={Object {}}
    type="text"
    value="d"
  />
</div>
`;

exports[`Pin Input should render correct style when size is provided 1`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 1.5rem;
  height: 1.5rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 0.75rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
</div>
`;

exports[`Pin Input should render correct style when size is provided 2`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 2rem;
  height: 2rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 0.875rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
</div>
`;

exports[`Pin Input should render correct style when size is provided 3`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 2.5rem;
  height: 2.5rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 1rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
</div>
`;

exports[`Pin Input should render correct style when size is provided 4`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 3rem;
  height: 3rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 1.125rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:invalid {
  border-color: rgb(220,53,69);
  box-shadow: rgb(220,53,69) 0px 0px 0px 1px;
  background-color: rgba(220,53,69,0.1);
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value=""
  />
</div>
`;

exports[`Pin Input should render correct when showState is false 1`] = `
.c0 {
  display: flex;
  align-items: center;
}

.c1 {
  width: 2.5rem;
  height: 2.5rem;
  margin-right: 0.375rem;
  outline: transparent solid 2px;
  outline-offset: 2px;
  font-size: 1rem;
  text-align: center;
  border-radius: 0.375rem;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(204,204,204);
  background-color: inherit;
  box-sizing: border-box;
}

.c1:focus {
  border-color: rgb(13,110,253);
  box-shadow: rgb(13,110,253) 0px 0px 0px 1px;
}

.c1:last-child {
  margin-right: 0;
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

<div
  className="c0"
  style={Object {}}
>
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={0}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="1"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={1}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="2"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={2}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="3"
  />
  <input
    aria-label="Please enter pin code"
    autoComplete="off"
    className="c1"
    data-index={3}
    inputMode="numeric"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="o"
    style={Object {}}
    type="text"
    value="4"
  />
</div>
`;
